### PR TITLE
Linear Regression Model and gRPC Support

### PR DESCRIPTION
**Pull request by inf0rmatiker**
## Notes

Sorry about the number of commits. Lots of testing going on here and pushing/pulling from home workstation to lattice cluster.

### CODE REVIEW

**Comment by drg101** [comment]
Can't comment on functionality, but can you add to the model_catalogue collection on mongo DB with whatever is relevant to this?

File: src/main/java/org/sustain/modeling/SustainLinearRegression.java [hunk 1]
```diff
@@ -0,0 +1,451 @@
+/* ---------------------------------------------------------------------------------------------------------------------
+ * LinearRegressionModel.java -
+ *      Defines a generalized linear regression model that can be
+ *      built and executed over a set of MongoDB documents.
+ *
+ * Author: Caleb Carlson
+ *
+ * LEGAL/SOFTWARE LICENSE STATEMENT
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/modeling/SustainLinearRegression.java:8]: This has got to go

File: src/main/java/org/sustain/modeling/SustainLinearRegression.java [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+import org.apache.logging.log4j.Logger;
+import scala.collection.Seq;
+
+import java.util.*;
+
+/**
+ * Provides an interface for building generalized Linear Regression
+ * models on data pulled in using Mongo's Spark Connector.
+ */
+public class SustainLinearRegression {
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/modeling/SustainLinearRegression.java:156]: These models should all `implement` a defined model `interface` so we are all on the same page about how a model implementation looks, and also to cut down on code duplication.

File: src/main/java/org/sustain/modeling/SustainLinearRegression.java [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+                .getOrCreate();
+
+        sparkContext = new JavaSparkContext(sparkSession.sparkContext());
+        addClusterDependencyJars();
+    }
+
+    /**
+     * Adds required dependency jars to the Spark Context member.
+     */
+    private void addClusterDependencyJars() {
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/modeling/SustainLinearRegression.java:229]: Dependency JARs should probably already exist on all the worker nodes pre-runtime, so we don't choke the network every time we launch a model. (However, after the first time uploading the JARs, Spark sees that they all exist and refuses to send them over again)

File: src/main/java/org/sustain/modeling/SustainLinearRegression.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+    /**
+     * Converts a Java List<String> of inputs to a Scala Seq<String>
+     * @param inputList The Java List<String> we wish to transform
+     * @return A Scala Seq<String> representing the original input list
+     */
+    public Seq<String> convertListToSeq(List<String> inputList) {
+        return JavaConverters.asScalaIteratorConverter(inputList.iterator()).asScala().toSeq();
+    }
+
+    public void buildAndRunModel() {
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/modeling/SustainLinearRegression.java:346]: This needs to be optimized, separate ticket should be added for that.

**Comment by inf0rmatiker** [comment]
Build is failing because proto needs to be regenerated.

File: Makefile [hunk 5]
```diff
@@ -1,10 +1,13 @@
+
 # --------------------------------------------------------------------
 # Author: Menuka Warushavithana
 # --------------------------------------------------------------------
 
 .EXPORT_ALL_VARIABLES:
 
 SERVER_HOST = lattice-165
+SERVER_PORT = 50051
```
Comment (inf0rmatiker) [inline Makefile:9]: No reason to have the server port hardcoded in the source code. We may want to use a different port in the future (30000-ish) if we use a K8s Service.

File: build.gradle [hunk 6]
```diff
@@ -28,31 +30,35 @@ dependencies {
...
     testCompile "org.mockito:mockito-core:1.9.5"
 
     testImplementation 'org.junit.jupiter:junit-jupiter-api:5.3.1'
     testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.3.1'
 }
 
-test {
-    useJUnitPlatform()
+
+task copyDependencies(type: Copy) {
```
Comment (inf0rmatiker) [inline build.gradle:55]: Copies the spark/mongo dependency JARs from the classpath to the `./build/libs` directory at build time so we can upload them to the Spark cluster at runtime.

File: src/main/java/org/sustain/db/mongodb/DBConnection.java [hunk 7]
```diff
@@ -25,7 +25,7 @@ public static MongoDatabase getConnection() {
         if (mongoClient == null) {
             log.info("Creating new connection to MongoDB");
             initDatabaseProperties();
-            mongoClient = new MongoClient(host, Integer.parseInt(port));
```
Comment (inf0rmatiker) [inline src/main/java/org/sustain/db/mongodb/DBConnection.java:28]: `PORT` gets parsed as an `Integer` from the environment variables now, so is already an `int`.

**Review by menuka94 — APPROVED** [review]

**Pull request merged**